### ISSUE TITLE
Wrong size of multiclass predictions when n_classes == 2

### ISSUE THREAD

**Issue by dreamflyer** [microsoft/LightGBM#2278]
Operating System: Mac OS X 10.10.55

CPU/GPU model: 2.2 GHz Intel Core i7, Intel Iris Pro 1536 MB

C++/Python/R version: 4.2.1/3.7.0/...

LightGBM version or commit hash: 7189743d224a405bb48050834460d107fb7c1c0b

```
model = lightgbm.LGBMModel(objective = "multiclass", n_classes=2)

model.fit(train_x, train_y, eval_set=[(val_x, val_y)], callbacks = [custom_callback], eval_metric = custom_metric)

model.predict(features) # lets say it contains features of ['dog', 'cat', 'dog', 'cat']
```

and i get something like [0.2, 0.5, 0.7, 0.1] (and same for y_pred in eval_metric callable, but according to doc it should be n_classes * input_length sized array). How should i interpret this result? Is it probability of 1st class or second?? If it is probability of first class then there is inconsistency since for [1, 0, 1, 0] ground truth ideal result will be [0, 1, 0, 1].

for n_classes > 2 everything is working as expected

**Comment by dreamflyer** [comment]
also, i found that for n_classes == 2 model.predict() returns same values as (1 - y_pred) in custom_metric callable

**Comment by guolinke** [comment]
n_classees == 2 will automatically convert the binary classification, which only produces the output of positive class.

**Comment by dreamflyer** [comment]
@guolinke so which of two classes is "positive"? Anyway this is inconsistent and unexpected behaviour. And what about differences between y_pred in custom metric callable and result of predict()?

**Comment by guolinke** [comment]
okay, I find the problem.
Instead of `LGBMModel`, you should use `LGBMClassifier`, which will automatically wrap the predict results for binary classes.

https://github.com/microsoft/LightGBM/blob/7189743d224a405bb48050834460d107fb7c1c0b/python-package/lightgbm/sklearn.py#L803-L860

**Comment by dreamflyer** [comment]
@guolinke LGBMModel works fine for all cases except of when n_classes =2, both for regression and classification. I use lightgbm for some pipeline where configuration of LGBMModel is calculated from some metadata. Basically you suggest that i should split my logic on two branches where i should detect what to use (LGBMClassifier or LGBMRegressor) which actually "generic "LGBMModel should do, according to docs. But it looks like some workaround and not solution.

**Comment by guolinke** [comment]
The implementation LGBMModel is to avoid the duplicate codes in LGBMClassifier, LGBMRegressor and LGBMRanker. Its purpose is no a class for all purposes. 

For example, you will need to pack `group` information in LGBMRanker, and need predicted class in LGBMClassifier. LGBMModel, as a base class, does not provide these function. 

And refer to sklearn's interface, the `predict` function of Classifier should return the predicted class, not the prob. Therefore, the predict and predict_prob is implemented in LGBMClassifier.

However, the document of  `LGBMModel.predict` needs to be clarified, for binary classification, it returns  [n_samples], not [n_samples, n_classes]. And the return values are the prob of the positive class (the label value greater than zero). 

@StrikerRUS could you help to update the related documentation?

**Comment by StrikerRUS** [comment]
@guolinke 
> could you help to update the related documentation?

Sure! How would you like to update it? It already says that shape can be `n_samples`.
https://github.com/microsoft/LightGBM/blob/86a957830b7caf2cedb7aaa23e824df6b38f7c7f/python-package/lightgbm/sklearn.py#L640-L641

**Comment by dreamflyer** [comment]
why it cannot be always just [n_samples, n_classes]? Even when n_classes = 1? thus you shouldn't always check format changes. right know i forced to handle 3 different cases when n_classes =1, n_classes = 2 and n_classes > 2, for what?? what is purpose of that? it is very annoying. Further more why input ground truth labels are like [1, 3, 2] instead of onehot encoding(while output predictions is onehot or probabilities EXCEPT of when n_classes is 1 or 2, and custom metric's y_pred is 1d [samples*classes] array). So in your model u have 3 **different** representations of **same** kind of data: input ground truth labels, y_pred in custom metric and predictions...and last two can change their format depends on number of classes...and you should over and over convert data before and after your model applying to make it standard and sane like in Keras or Thensorflow

**Comment by guolinke** [comment]
firstly, only returning the probability of positive class is commonly used in binary classification, since we only need this and the probability of negative class could be easily restored. And you can use LGBMClassifier as stated above, not the `LGBMModel`, which is not designed for the classification problem.
Secondly, what is the case of n_class=1? 

as for the one-hot coding label format, I don't think there is a standard. NN toolkits use this format for it is more convenient to back-prop loss. LightGBM uses int label for its memory-efficient.

In short, for these wrapping for multi-class outputs, you can simply write some functions and handle them anywhere.
you can also use the high-level multi-class wrappers, such as https://scikit-learn.org/stable/modules/generated/sklearn.multiclass.OneVsRestClassifier.html, and use LightGBM as its binary classifier.
